cpp/src/fil/infer.cu

```diff
@@ -248,20 +248,33 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     __syncthreads();  // free up input row
     // load margin into shared memory
     auto per_thread = (vec<NITEMS, float>*)tmp_storage;
-    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+    per_thread[threadIdx.x] = acc;
 
     __syncthreads();
     // reduce per-thread margin summand into per-class complete margin
     // (for each of the NITEMS rows)
-    // TODO(levsnv): use CUB/tree reduction when num_classes is small
-    if (threadIdx.x < num_classes) {
-      for (int c = threadIdx.x + num_classes; c < blockDim.x; c += num_classes)
-        acc += per_thread[c];
+    int t = threadIdx.x;
+    int sets = blockDim.x / num_classes;
+    for (; sets > 3; sets = sets / 4 + sets % 4) {
+      if (t < sets / 4 * num_classes) {
+        int add_start = t + sets % 4 * num_classes;
+        per_thread[t] = acc =
```

Comment:
Will the code become easier to understand if 4 is replaced with `N`, where `N` is a compile-time constant? The sum can then be expressed as a loop with `#pragma unroll`.

Reply:
I can try. I will add comments explaining the general idea behind this as well